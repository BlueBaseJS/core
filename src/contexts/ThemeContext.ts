import { Configs } from '../Configs';
import { Theme } from '..';
import { createContext } from 'react';

/**
 * Interface of object passed as param to the ThemeConsumer render prop method.
 */
export interface ThemeContextData {
	/** Helper method to change current theme. */
	changeTheme: (slug: string) => void;

	changeMode: (mode: Configs['theme.mode']) => void;

	/** Selected theme varaint */
	theme: Theme;
}

/**
 * This is the context where BlueBase theme is stored.
 */
export const ThemeContext: React.Context<ThemeContextData> = createContext(undefined as any);

/**
 * BlueBase theme consumer.
 */
export const ThemeConsumer = ThemeContext.Consumer;
