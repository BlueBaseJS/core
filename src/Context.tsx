import { BlueBase } from './BlueBase';
import { createContext } from 'react';

export const BlueBaseContext: React.Context<BlueBase> = createContext(undefined as any);
export const BlueBaseProvider = BlueBaseContext.Provider;
export const BlueBaseConsumer = BlueBaseContext.Consumer;
