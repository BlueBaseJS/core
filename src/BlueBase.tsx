import { Analytics, Logger } from './api';
import {
	ComponentCollection,
	ComponentRegistry,
	ConfigCollection,
	ConfigRegistry,
	HookNestedCollection,
	HookRegistry,
	PluginCollection,
	PluginRegistry,
	ThemeCollection,
	ThemeRegistry,
} from './registries';
import { BlueBaseProvider } from './Context';
import React from 'react';
import { ThemeProvider } from './themes';
import { getComponent } from './getComponent';
import systemHooks from './hooks';

export interface BootOptions {
	components: ComponentCollection,
	configs: ConfigCollection,
	hooks: HookNestedCollection,
	plugins: PluginCollection,
	themes: ThemeCollection,
	// routes: Plugin[]
}

export class BlueBase {

	// APIs
	public Analytics = new Analytics(this);
	public Logger = new Logger(this);

	// Registries
	public Components = new ComponentRegistry(this);
	public Configs = new ConfigRegistry(this);
	public Hooks = new HookRegistry(this);
	public Plugins = new PluginRegistry(this);
	public Themes = new ThemeRegistry(this);

	// Flags
	public booted = false;

	private bootOptions: BootOptions = {
		components: {},
		configs: {},
		hooks: {},
		plugins: [],
		themes: [],
	};

	public async boot(options?: Partial<BootOptions> & { children?: React.ReactNode }) {

		this.bootOptions = { ...this.bootOptions, ...options };

		await this.Hooks.registerNestedCollection(systemHooks);
		await this.Hooks.registerNestedCollection(this.bootOptions.hooks);

		await this.Hooks.run('bluebase.boot', this.bootOptions);

		const SystemApp = getComponent('SystemApp');

		// Set View
		// const SystemApp = this.Components.resolve('SystemApp');
		// SystemApp = await this.Hooks.run('bluebase.system.app', SystemApp);

		const BlueBaseRoot = () => (
			<BlueBaseProvider value={this}>
				<ThemeProvider>
					<SystemApp>
						{options && options.children}
					</SystemApp>
				</ThemeProvider>
			</BlueBaseProvider>
		);

		this.booted = true;

		return BlueBaseRoot;
	}
}