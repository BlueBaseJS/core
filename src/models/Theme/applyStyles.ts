import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import { ComponentStyles } from './Theme';
import React from 'react';
import deepmerge from 'deepmerge';

export type ComponentWithDefaultStyles = React.ComponentType<any> & { defaultStyles?: ComponentStyles };

/**
 * Merges component styles from different sources and passes on the the component as 'styles' prop.
 *
 * Following sources are considered as style sources. (In order of least important first):
 *
 * 1. defaultStyles: A component may provide default styles as a "defaultStyles" static
 * property of component. This is similar to defaultProps in react.
 * 2. stylesParam: 3rd property of this function. Note: Component Registry passes 'styles' prop from it's registry item.
 * 3. Themes: From theme.components[componentName] property of current theme.
 * 4. styles prop: The styles prop passed on to the component during usage.
 *
 * @param name
 * @param Component
 * @param stylesParam
 */
export const applyStyles = (name: string, Component: ComponentWithDefaultStyles, stylesParam: ComponentStyles = {}) =>
{

	return class ThemedComponent extends React.PureComponent<any> {

		static contextType = BlueBaseContext;

		render() {
			const BB: BlueBase = (this as any).context;
			const { styles: stylesProp = {}, ...rest } = this.props;

			const defaultStyles = Component.defaultStyles || {};

			const theme = BB.Themes.get(BB.Configs.getValue('theme'));
			const themedStyles = (theme && theme.components && theme.components[name]) ? theme.components[name] : {};

			const styles = deepmerge.all([
				defaultStyles,
				stylesParam,
				themedStyles,
				stylesProp
			]);

			return React.createElement(Component, { ...rest, styles });
		}
	};
};