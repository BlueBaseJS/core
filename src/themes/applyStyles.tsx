import { ComponentStyles } from './Theme';
import { MaybeThunk } from '../utils';
import React from 'react';
import { useStyles } from '../hooks';

export type BlueBaseComponent<T = any> = React.ComponentType<T> & {
	defaultStyles?: ComponentStyles;
};

export type ComponentWithDefaultStyles = BlueBaseComponent;

export interface ThemedComponentProps {
	styles: ComponentStyles;

	[key: string]: any;
}

export type applyStylesOptions =
	| string
	| {
			name: string;

			/**
			 * @deprecated
			 */
			styles?: MaybeThunk<ComponentStyles>;
	  };

/**
 * Merges component styles from different sources and passes on the the component as 'styles' prop.
 *
 * Following sources are considered as style sources. (In order of least important first):
 *
 * 1. defaultStyles: A component may provide default styles as a "defaultStyles" static
 * property of component. This is similar to defaultProps in react.
 * 2. stylesParam: 3rd property of this function. Note: Component Registry passes 'styles' prop from it's registry item.
 * 3. Themes: From theme.components[componentName] property of current theme.
 * 4. styles prop: The styles prop passed on to the component during usage.
 *
 * @param options
 */

export function applyStyles<T = any>(options: applyStylesOptions) {
	return (Component: BlueBaseComponent<T>): BlueBaseComponent<T> => {
		const name = typeof options === 'string' ? options : options.name;

		const ThemedComponent = (props: T) => {
			// const { styles: stylesProp, ...rest } = props;
			const styles = useStyles(name, props, Component.defaultStyles);

			// If we were able to extract any rules, pass them forward in the styles prop
			const finalProps = Object.keys(styles).length > 0 ? { ...props, styles } : props;

			return React.createElement(Component, finalProps);
		};

		ThemedComponent.displayName = 'ThemedComponent';

		return ThemedComponent;
	};
}
