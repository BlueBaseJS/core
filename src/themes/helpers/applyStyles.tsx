import { ComponentStyles, ThemeValue } from '../structure/Theme';
import { MaybeThunk, resolveThunk } from '../../utils';
import React from 'react';
import { ThemeContext } from '..';
import { buildTheme } from './buildTheme';
import deepmerge from 'deepmerge';
import isNil from 'lodash.isnil';

export type ComponentWithDefaultStyles = React.ComponentType<any> & { defaultStyles?: ComponentStyles };

export interface ThemedComponentProps {
	styles: ComponentStyles,

	[key: string]: any,
}

/**
 * Merges component styles from different sources and passes on the the component as 'styles' prop.
 *
 * Following sources are considered as style sources. (In order of least important first):
 *
 * 1. defaultStyles: A component may provide default styles as a "defaultStyles" static
 * property of component. This is similar to defaultProps in react.
 * 2. stylesParam: 3rd property of this function. Note: Component Registry passes 'styles' prop from it's registry item.
 * 3. Themes: From theme.components[componentName] property of current theme.
 * 4. styles prop: The styles prop passed on to the component during usage.
 *
 * FIXME: Fix return typing of this function
 *
 * @param name
 * @param Component
 * @param stylesParam
 */
export const applyStyles = (
	{ name, styles: stylesParam }: {
		name?: string,
		styles?: MaybeThunk<ComponentStyles>
	} = {}
) => (Component: ComponentWithDefaultStyles)
: React.ComponentType<any> =>
{

	return class ThemedComponent extends React.Component<ThemedComponentProps> {

		static contextType = ThemeContext;

		render() {

			const context: { theme: ThemeValue } = this.context;
			const { styles: stylesProp, ...rest } = this.props;

			// Extract defaultStyles from component
			const defaultStyles = Component.defaultStyles;

			// Extract theme, or use default theme as placeholder
			// We need to do this, because when the actual theme is being resolved,
			// we may need to show Loading state. That state may need some styles as well.
			const theme = context ? context.theme : buildTheme()();

			// Extract component style rules from theme
			let themedStyles;
			if (name) {
				const themeComponentStyles = theme.components;

				// Extract style rules for this component
				themedStyles = themeComponentStyles[name];
			}

			// Put all style rules in an array
			const stylesArr: any[] = [
				defaultStyles,
				stylesParam,
				themedStyles,
				stylesProp,
			]
			// Remove those styles which are nil
			.filter(a => !isNil(a))
			// If any item is a thunk, resolve it
			.map(a => a && resolveThunk(a, theme));

			// Merge all into a single object
			const styles = deepmerge.all(stylesArr);

			// If we were able to extract any rules, pass them forward in the styles prop
			const props = stylesArr.length > 0 ? { ...rest, styles } : rest;

			return React.createElement(Component, props);
		}
	};
};