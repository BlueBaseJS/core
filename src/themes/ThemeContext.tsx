import React, { createContext, useContext, useEffect, useState } from 'react';

import { BlueBase } from '../BlueBase';
import { BlueBaseContext } from '../Context';
import { Configs } from '../Configs';
import { Theme } from '../registries';
import { ThemeValueInput } from './structure';
import { buildTheme } from './helpers';
import deepmerge from 'deepmerge';
import { useConfig } from '../hooks';

/**
 * Props of the `ThemeProvider` component.
 */
export interface ThemeProviderProps {
	/**
	 * Key of the theme to use for children components. If this prop is not set,
	 * the globally selected theme is used.
	 */
	theme?: Configs['theme.name'];

	/**
	 * Theme mode
	 */
	mode?: Configs['theme.mode'];

	/**
	 * Any custom overrides to the selected theme.
	 */
	overrides?: ThemeValueInput;

	children: React.ReactNode;
}

/**
 * Interface of object passed as param to the ThemeConsumer render prop method.
 */
export interface ThemeContextData {
	/** Helper method to change current theme. */
	changeTheme: (slug: string) => void;

	/** Current theme */
	theme: Theme;
}

/**
 * This is the context where BlueBase theme is stored.
 */
export const ThemeContext: React.Context<ThemeContextData> = createContext(undefined as any);

/**
 * BlueBase theme consumer.
 */
export const ThemeConsumer = ThemeContext.Consumer;

/**
 * ðŸŽ¨ ThemeProvider
 */
export const ThemeProvider = (props: ThemeProviderProps) => {
	const BB: BlueBase = useContext(BlueBaseContext);

	const [themeName, setThemeName] = useConfig('theme.name');
	const [modeConfig] = useConfig('theme.mode');
	const [overridesConfig] = useConfig('theme.overrides');

	const name = props.theme || themeName;
	const mode = props.mode || modeConfig;
	const overrides = deepmerge.all([{}, overridesConfig || {}, props.overrides || {}]);

	const DEFAULT_THEME = buildTheme(mode)({ value: overrides });

	const [theme, setThemeState] = useState<Theme>(DEFAULT_THEME);

	/**
	 * Sets a theme to Provider's state. If a theme key is given, it is used,
	 * otherwise global theme is used.
	 *
	 * @param slug
	 * @param overrides
	 */
	async function setTheme(slug: string, cancelled: boolean) {
		try {
			const resolvedTheme = await BB.Themes.resolve(slug);
			const modeTheme =
				resolvedTheme.mode !== mode ? await BB.Themes.resolveAlternate(resolvedTheme.key) : {};

			const change = deepmerge.all([resolvedTheme, modeTheme, overrides]) as Theme;

			if (JSON.stringify(theme) === JSON.stringify(change)) {
				return;
			}

			if (!cancelled) {
				setThemeState(change);
			}
		} catch (error) {
			BB.Logger.warn(
				`Could not change theme. Reason: Theme with the key "${slug}" does not exist.`
			);
		}
	}

	useEffect(() => {
		if (theme && theme.key === name) {
			return;
		}

		let cancelled = false;
		setTheme(name, cancelled);

		return () => {
			cancelled = true;
		};
	}, [name, mode, overrides]);

	const value: ThemeContextData = {
		changeTheme: setThemeName,
		theme,
	};

	return <ThemeContext.Provider value={value} children={props.children} />;
};

ThemeProvider.displayName = 'ThemeProvider';
