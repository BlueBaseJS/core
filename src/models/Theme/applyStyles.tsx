import { ComponentStyles, ThemeValue } from './Theme';
import { MaybeThunk, resolveThunk } from '../../utils';
import React from 'react';
import { ThemeContext } from '../../themes';
import deepmerge from 'deepmerge';
import isNil from 'lodash.isnil';

export type ComponentWithDefaultStyles = React.ComponentType<any> & { defaultStyles?: ComponentStyles };

export interface ThemedComponentProps {
	styles: ComponentStyles,

	[key: string]: any,
}

/**
 * Merges component styles from different sources and passes on the the component as 'styles' prop.
 *
 * Following sources are considered as style sources. (In order of least important first):
 *
 * 1. defaultStyles: A component may provide default styles as a "defaultStyles" static
 * property of component. This is similar to defaultProps in react.
 * 2. stylesParam: 3rd property of this function. Note: Component Registry passes 'styles' prop from it's registry item.
 * 3. Themes: From theme.components[componentName] property of current theme.
 * 4. styles prop: The styles prop passed on to the component during usage.
 *
 * FIXME: Fix return typing of this function
 *
 * @param name
 * @param Component
 * @param stylesParam
 */
export const applyStyles = (
	name: string,
	Component: ComponentWithDefaultStyles,
	stylesParam: MaybeThunk<ComponentStyles> = {})
: React.ComponentType<any> =>
{

	return class ThemedComponent extends React.PureComponent<ThemedComponentProps> {

		render() {

			const { styles: stylesProp, ...rest } = this.props;

			return (
				<ThemeContext.Consumer children={(args?: { theme: ThemeValue }) => {

					if(!args) {
						return;
					}

					const { theme } = args;

					const defaultStyles = Component.defaultStyles;
					const themedStyles = theme.components[name];

					const stylesArr: any = [
						defaultStyles,
						stylesParam,
						themedStyles,
						stylesProp,
					]
					.filter(a => !isNil(a))
					.map(a => a && resolveThunk(a, theme));

					const styles = deepmerge.all(stylesArr);

					return React.createElement(Component, { ...rest, styles });
				}} />
			);
		}
	};
};