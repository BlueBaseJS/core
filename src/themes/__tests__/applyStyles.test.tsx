import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import React from 'react';
import { Text } from 'react-native';
import { ThemeProvider } from '../ThemeContext';
import { applyStyles } from '../applyStyles';
import { mount } from 'enzyme';
import { waitForElement } from 'enzyme-async-helpers';

const Demo = (props: any) => <Text style={props.styles.root}>Demo Text</Text>;
Demo.defaultStyles = { root: { backgroundColor: 'red' } };

describe('applyStyles', () => {
	test(`apply styles without theme & unknown component name`, async () => {
		const ThemedDemo = applyStyles({ name: 'Foo' })(Demo);

		const BB = new BlueBase();
		await BB.boot();

		const wrapper = mount(
			<BlueBaseContext.Provider value={BB}>
				<ThemeProvider>
					<ThemedDemo />
				</ThemeProvider>
			</BlueBaseContext.Provider>
		);

		// Wait for render
		await waitForElement(wrapper as any, ThemedDemo);

		const style =
			wrapper
				.find('Demo Text')
				.first()
				.prop('style') || {};
		expect(style.backgroundColor).toBe('red');
	});
});
