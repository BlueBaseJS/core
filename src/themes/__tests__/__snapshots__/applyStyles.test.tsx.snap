// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyStyles apply styles without theme & unknown component name 1`] = `
<ThemedComponent>
  <Demo
    styles={
      Object {
        "root": Object {
          "backgroundColor": "red",
        },
      }
    }
  >
    <Text
      style={
        Object {
          "backgroundColor": "red",
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        Demo Text
      </Text>
    </Text>
  </Demo>
</ThemedComponent>
`;

exports[`applyStyles apply styles without theme 1`] = `
<ThemedComponent>
  <Demo
    styles={
      Object {
        "root": Object {
          "backgroundColor": "red",
        },
      }
    }
  >
    <Text
      style={
        Object {
          "backgroundColor": "red",
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        Demo Text
      </Text>
    </Text>
  </Demo>
</ThemedComponent>
`;
